geocode_array/geocode_array.py

```diff
@@ -0,0 +1,98 @@
+# -*- coding: utf-8 -*-
+"""
+CCT Geocode Project:
+This code reads in a list of addresses in json or csv format, attempts to geocode these addresses with the ArcGIS and Nominatim geocoders,
+compares the results from both geocoders and selects the best. Then it attempts to geocode the addresses with the 
+CCT geocoder and returns the output from the CCT geocoder as well as the best result from the combination of ArcGIS and Nominatim
+in json and csv format.
+"""
+
+import logging
+import itertools
+
+from geocode_array import DISPERSION_THRESHOLD
+
+
+def _get_geocoders(result_tuples) -> [str]:
+    gc_classes = [
+        result_tuple[0] for result_tuple in result_tuples
+    ]
+
+    return gc_classes
+
+
+def _find_mean_pos(result_tuples) -> float:
+    lat_sum = sum([result_tuple[2] for result_tuple in result_tuples])
+    lon_sum = sum([result_tuple[3] for result_tuple in result_tuples])
+    tuple_len = len(result_tuples)
+
+    lat_mean = lat_sum / tuple_len
+    lon_mean = lon_sum / tuple_len
+
+    return lat_mean, lon_mean
+
+
+def _find_dist_mean(lat_mean, lon_mean, result_tuples) -> float:
+    """Find the average Euclidean distance to a point"""
+    dist_sum = sum([
+        ((result_tuple[2] - lat_mean) ** 2 + (result_tuple[3] - lon_mean) ** 2) ** 0.5
+        for result_tuple in result_tuples
+    ])
+    tuple_len = len(result_tuples)
+
+    return dist_sum / tuple_len
+
+
+def combine_results(result_tuples) -> (float or None, float or None, float or None, [str] or None):
+    """
+    Compares outputs from various geocoders. Assumes that there are no null values
+
+    Uses the following algorithm:
+        1. Finds the average location for the combination of results that has the lowest dispersion (mean distance to
+        centre, within a certain threshold (~one hectare). Starts out with trying to combine everything, then work
+        through the permutations.
+        2. If there is no combination that has a dispersion below that threshold, select the result with the lowest
+        internal error
+        3. Finally, if that isn't possible, just take the first value
+    """
+    # Finding the average location
+    logging.debug(f"Trying all results from {_get_geocoders(result_tuples)}...")
+    lat_mean, lon_mean = _find_mean_pos(result_tuples)
+    dispersion = _find_dist_mean(lat_mean, lon_mean, result_tuples)
+
+    if dispersion < DISPERSION_THRESHOLD:
+        logging.debug("Returning mean of all results")
+        return lat_mean, lon_mean, dispersion, _get_geocoders(result_tuples)
+
+    # Working through various permutations
+    for c in range(len(result_tuples) - 1, 1, -1):
+        logging.debug(f"Taking {c} values...")
+        perm_means = [
+            _find_mean_pos(perm_tuples)
+            for perm_tuples in itertools.permutations(result_tuples, c)
```

Comment:
itertools combinations might be more efficient in this case, and will still allow for a setting of group size

Reply:
Ah, you're right, I do want combinations not permutations - because it's a linear calculation, I only care about the unique combinations of elements, not their sorting within the combination (which is what permutations does).

Will change.